src/player.cpp

```diff
@@ -8403,179 +8403,217 @@ void player::vomit()
-            d_start /= 3;
-        }
-    } else {
-        if (has_trait("SLIMY")) {
-            dur *= 1.2;
-            d_start *= 1.2;
-        }
-    }
-
-    add_morale(MORALE_WET, morale_effect, morale_cap, dur, d_start);
-}
-
-void player::drench_mut_calc()
-{
-    mMutDrench.clear();
-    int ignored, neutral, good;
-
-    for( auto &elem : mDrenchEffect ) {
-        ignored = 0;
-        neutral = 0;
-        good = 0;
-
-        for( const auto &_iter : my_mutations ) {
-            const auto &mdata = mutation_branch::get( _iter.first );
-            const auto _wp_iter = mdata.protection.find( elem.first );
-            if( _wp_iter != mdata.protection.end() ) {
-                ignored += _wp_iter->second.x;
-                neutral += _wp_iter->second.y;
-                good += _wp_iter->second.z;
-            }
-        }
-
-        mMutDrench[elem.first]["good"] = good;
-        mMutDrench[elem.first]["neutral"] = neutral;
-        mMutDrench[elem.first]["ignored"] = ignored;
-    }
+    add_morale( MORALE_WET, morale_effect, total_morale, 5, 5, true );
 }
 
-void player::update_body_wetness()
+void player::update_body_wetness( const w_point &weather )
 {
-    /**
-    * Issue : morale and wetness still aren't linked ... they are two seperate things that mostly coincide
-    **/
+    // Average number of turns to go from completely soaked to fully dry
+    // assuming average temperature and humidity
+    constexpr int average_drying = HOURS(1);
 
-    /*
-    * Mutations and weather can affect the duration of the player being wet.
-    */
-    int delay = 10;
+    // A modifier on drying time
+    double delay = 1.0;
+    // Weather slows down drying
+    delay -= ( weather.temperature - 65 ) / 100.0;
+    delay += ( weather.humidity - 66 ) / 100.0;
+    delay = std::max( 0.1, delay );
+    // Fur/slime retains moisture
     if( has_trait("LIGHTFUR") || has_trait("FUR") || has_trait("FELINE_FUR") ||
         has_trait("LUPINE_FUR") || has_trait("CHITIN_FUR") || has_trait("CHITIN_FUR2") ||
         has_trait("CHITIN_FUR3")) {
-        delay += 2;
-    }
-    if (has_trait("URSINE_FUR")) {
-        delay += 5;
+        delay = delay * 6 / 5;
     }
-    if (has_trait("SLIMY")) {
-        delay -= 5;
+    if( has_trait( "URSINE_FUR" ) || has_trait( "SLIMY" ) ) {
+        delay = delay * 3 / 2;
     }
-    if (g->weather == WEATHER_SUNNY) {
-        delay -= 2;
+
+    if( !one_in_improved( average_drying * delay / 100.0 ) ) {
+        // No drying this turn
+        return;
     }
 
-    if (!calendar::once_every(delay)) {
+    // Now per-body-part stuff
+    // To make drying uniform, make just one roll and reuse it
+    const int drying_roll = rng( 1, 100 );
+    if( drying_roll > 40 ) {
+        // Wouldn't affect anything
         return;
     }
 
-    for (int i = 0; i < num_bp; ++i) {
-        if (body_wetness[i] == 0) {
+    for( int i = 0; i < num_bp; ++i ) {
+        const body_part bp = static_cast<body_part>( i );
```

Comment:
It might be useful to have a global array of all the body parts, so one can write
```C++
for( const body_part bp : all_body_parts ) {
```
And won't need the `num_bp` nor the static_cast.